torch/csrc/autograd/python_function.cpp

```diff
@@ -125,14 +131,13 @@ static int THPFunction_clear(THPFunction *self)
   self->saved_variables = NULL;
   delete saved_variables;
 
-  auto output_backward_hooks = self->output_backward_hooks;
-  self->output_backward_hooks = NULL;
-  delete[] output_backward_hooks;
-
   auto output_info = self->output_info;
   self->output_info = NULL;
   delete output_info;
 
+  self->cdata.pre_hooks.clear();
```

Comment:
Doesn't this need something like `Py_CLEAR`? I'm afraid that in the middle of the `clear()` call, the GC might kick in, and will call traverse. I think it would be safer to swap with a local empty vector (no need to call `clear()`, because it will be destructed just a moment later).